### PR TITLE
RFC 89: SQL query logging callback

### PR DESCRIPTION
**Pull request by elpaso**
Implementation of RFC89: sql logging callback

### CODE REVIEW

File: ogr/ogrsf_frmts/sqlite/ogrsqlitedatasource.cpp [hunk 1]
```diff
@@ -4664,6 +4680,53 @@ void OGRSQLiteBaseDataSource::SetEnvelopeForSQL(const CPLString &osSQL,
...
+/*                       SetQueryLoggerFunc()                          */
+/***********************************************************************/
+
+bool OGRSQLiteBaseDataSource::SetQueryLoggerFunc(
+    GDALQueryLoggerFunc pfnQueryLoggerFuncIn, void *poQueryLoggerArgIn)
+{
+
+#if SQLITE_VERSION_NUMBER < 3014000
+    return false;
+#endif
```
Comment (rouault) [inline ogr/ogrsf_frmts/sqlite/ogrsqlitedatasource.cpp:4693]:
```suggestion
#else
```

File: ogr/ogrsf_frmts/sqlite/ogrsqlitedatasource.cpp [hunk 2]
```diff
@@ -4664,6 +4680,53 @@ void OGRSQLiteBaseDataSource::SetEnvelopeForSQL(const CPLString &osSQL,
...
+                                                   executionTimeMilliSeconds,
+                                                   source->poQueryLoggerArg);
+                    }
+                }
+                return 0;
+            },
+            reinterpret_cast<void *>(this));
+        return true;
+    }
+    return false;
```
Comment (rouault) [inline ogr/ogrsf_frmts/sqlite/ogrsqlitedatasource.cpp:4727]:
```suggestion
    return false;
#endif
```

File: ogr/ogrsf_frmts/sqlite/ogrsqlitedatasource.cpp [hunk 3]
```diff
@@ -4664,6 +4680,53 @@ void OGRSQLiteBaseDataSource::SetEnvelopeForSQL(const CPLString &osSQL,
...
+/***********************************************************************/
+/*                       SetQueryLoggerFunc()                          */
+/***********************************************************************/
+
+bool OGRSQLiteBaseDataSource::SetQueryLoggerFunc(
+    GDALQueryLoggerFunc pfnQueryLoggerFuncIn, void *poQueryLoggerArgIn)
+{
+
+#if SQLITE_VERSION_NUMBER < 3014000
+    return false;
```
Comment (rouault) [inline ogr/ogrsf_frmts/sqlite/ogrsqlitedatasource.cpp:4692]:
```suggestion
    (void)pfnQueryLoggerFuncIn;
    (void)poQueryLoggerArgIn;
    return false;
```

File: ogr/ogrsf_frmts/sqlite/ogrsqlitebase.h [hunk 4]
```diff
@@ -197,6 +197,7 @@ class OGRSQLiteBaseDataSource CPL_NON_FINAL : public GDALPamDataset
     GetLayerWithGetSpatialWhereByName(const char *pszName) = 0;
 
     virtual OGRErr AbortSQL() override;
+    bool SetQueryLoggerFunc(GDALQueryLoggerFunc pfnQueryLoggerFuncIn, void*poQueryLoggerArgIn) override;
```
Comment (rouault) [inline ogr/ogrsf_frmts/sqlite/ogrsqlitebase.h:200]: format linter not happy here

File: autotest/cpp/test_ogr.cpp [hunk 5]
```diff
@@ -2213,4 +2215,108 @@ TEST_F(test_ogr, feature_defn_geomfields_iterator)
...
 
+// Test GDALDataset QueryLoggerFunc callback
+TEST_F(test_ogr, GDALDatasetSetQueryLoggerFunc)
+{
+    if (GDALGetDriverByName("GPKG") == nullptr)
+    {
+        GTEST_SKIP() << "GPKG driver missing";
+    }
+
+    auto tmpGPKG{testing::TempDir() + "/poly-1-feature.gpkg"};
```
Comment (rouault) [inline autotest/cpp/test_ogr.cpp:2226]: this test should also be skipped if the sqlite version if SQLITE_VERSION_NUMBER < 3014000  . You'll need to include sqlite3.h and add something like the following in autotest/cpp/CMakeLists.txt:
```
target_include_directories(gdal_unit_test PRIVATE $<TARGET_PROPERTY:SQLite::SQLite3,INTERFACE_INCLUDE_DIRECTORIES>)
```

File: ogr/ogrsf_frmts/sqlite/ogrsqlitebase.h (hunk 4 above)
Comment (elpaso) [reply ogr/ogrsf_frmts/sqlite/ogrsqlitebase.h:200]: yeah, I've fixed that but I'm stuck on ogr/ogrsf_frmts/sqlite/ogrsqlitedatasource.cpp:4695,style,unreachableCode,Statements following return, break, continue, goto or throw will never be executed.

Comment (rouault) [reply ogr/ogrsf_frmts/sqlite/ogrsqlitebase.h]: > yeah, I've fixed that but I'm stuck on ogr/ogrsf_frmts/sqlite/ogrsqlitedatasource.cpp:4695,style,unreachableCode,Statements following return, break, continue, goto or throw will never be executed.

See my suggestion of adding a #else / #endif that should hopefully fix it

**Pull request merged**